src/openrct2/actions/RideSetPriceAction.hpp

```diff
@@ -0,0 +1,190 @@
+#include "../localisation/Localisation.h"
+#include "../localisation/StringIds.h"
+#include "../management/Finance.h"
+#include "../ride/Ride.h"
+#include "../ride/RideData.h"
+#include "../ride/ShopItem.h"
+#include "../world/Park.h"
+#include "../world/Sprite.h"
+#include "GameAction.h"
+
+DEFINE_GAME_ACTION(RideSetPriceAction, GAME_COMMAND_SET_RIDE_PRICE, GameActionResult)
+{
+private:
+    NetworkRideId_t _rideIndex{ -1 };
+    money16 _price = MONEY16_UNDEFINED;
+    bool _primaryPrice = true;
+
+public:
+    RideSetPriceAction()
+    {
+    }
+    RideSetPriceAction(ride_id_t rideIndex, money16 price, bool primaryPrice)
+        : _rideIndex(rideIndex)
+        , _price(price)
+        , _primaryPrice(primaryPrice)
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags() | GA_FLAGS::ALLOW_WHILE_PAUSED;
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << DS_TAG(_rideIndex) << DS_TAG(_price);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        GameActionResult::Ptr res = std::make_unique<GameActionResult>();
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_PARK_RIDE_TICKETS;
+
+        if (_rideIndex >= MAX_RIDES || _rideIndex == RIDE_ID_NULL)
+        {
+            log_warning("Invalid game command for ride %u", _rideIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        Ride* ride = get_ride(_rideIndex);
+        if (ride == nullptr || ride->type == RIDE_TYPE_NULL)
+        {
+            log_warning("Invalid game command, ride_id = %u", _rideIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        rct_ride_entry* rideEntry = get_ride_entry(ride->subtype);
+        if (rideEntry == nullptr)
+        {
+            log_warning("Invalid game command for ride %u", _rideIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        return res;
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        GameActionResult::Ptr res = std::make_unique<GameActionResult>();
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_PARK_RIDE_TICKETS;
+
+        Ride* ride = get_ride(_rideIndex);
+        if (ride == nullptr || ride->type == RIDE_TYPE_NULL)
+        {
+            log_warning("Invalid game command, ride_id = %u", _rideIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        rct_ride_entry* rideEntry = get_ride_entry(ride->subtype);
+        if (rideEntry == nullptr)
+        {
+            log_warning("Invalid game command for ride %u", _rideIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        if (ride->overall_view.xy != RCT_XY8_UNDEFINED)
+        {
+            LocationXYZ16 coord;
+            coord.x = ride->overall_view.x * 32 + 16;
```

Comment:
set the `res->Position` coords as well

Reply:
I wanted to refactor those in another PR since other game actions are missing out here too.